src/hardware/driver/dac.rs

```diff
@@ -0,0 +1,240 @@
+            dac_code = DacCode::MAX_DAC_WORD - dac_code - 1; // For now just subtract a bit to stay in bounds for 0 A
+        }
+
+        if !(0..DacCode::MAX_DAC_WORD).contains(&dac_code) {
+            return Err(Error::Bounds);
+        };
+
+        Ok(Self(dac_code as u32))
+    }
+}
+
+impl From<DacCode> for u32 {
+    fn from(code: DacCode) -> u32 {
+        code.0
+    }
+}
+
+pub struct Dac<SPI: Transfer<u8> + Write<u8>> {
+    spi: SPI,
+    sync_n: gpio::ErasedPin<gpio::Output>,
+    channel: ChannelVariant,
+}
+
+impl<SPI, E> Dac<SPI>
+where
+    SPI: Transfer<u8, Error = E> + Write<u8, Error = E>,
+    E: Debug,
+{
+    /// Construct a new [Dac] in the following state:
+    /// - reset
+    /// - 10 V plusminus 1.25 V refernece span configured
+    /// - temperature drift calibration performed
+    /// - deglitch circuit is enabled by default (fast-settling off)
+    /// - asyncronous output update (when sync goes high) by default
+    /// - 0.5-MHz max DAC update rate by default for enhanced THD performance
+    /// - TNH masked for code jump > 2^14 (default)
+    pub fn new(
+        spi: SPI,
+        sync_n: gpio::ErasedPin<gpio::Output>,
+        channel: ChannelVariant,
+    ) -> Self {
+        let mut dac = Dac {
+            spi,
+            sync_n,
+            channel,
+        };
+
+        // reset DAC
+        dac.write(TRIGGER::SRST::RESET, DAC_ADDR::TRIGGER);
+        // set to 10 V plusminus 1.25 V referenece span
+        dac.write(CONFIG1::VREFVAL::SPAN_10V_1_25V, DAC_ADDR::CONFIG1);
+        // perform calibration
+        dac.calibrate();
+
+        dac
+    }
+
+    /// Perform temperature drift calibration (waits until calibration is done).
+    pub fn calibrate(&mut self) {
+        self.write(CONFIG1::EN_TMP_CAL::ENABLED, DAC_ADDR::CONFIG1);
+        self.write(TRIGGER::RCLTMP::RECAL, DAC_ADDR::TRIGGER);
+        // continuously read recalibration done bit until it is set
+        while (self.read(DAC_ADDR::STATUS) & STATUS::ALM::RECALIBRATED) == 0 {}
+    }
+
+    /// Set the DAC to produce a voltage corresponding to `current`.
+    pub fn set(&mut self, current: f32) -> Result<(), Error> {
+        let dac_code = DacCode::try_from((current, self.channel))?;
+        self.write(dac_code.0 << 4, DAC_ADDR::DAC_DATA);
+        Ok(())
+    }
+
+    /// Write 24 bit [data] to the DAC at [addr].
+    pub fn write(&mut self, data: u32, addr: DAC_ADDR) {
+        let bytes = (addr as u32 | (data & 0xffffff)).to_be_bytes();
+        self.sync_n.set_low();
+        self.spi.write(&bytes).unwrap();
+        self.sync_n.set_high();
+    }
+
+    /// Perform a read of a DAC register. First writes a word and then reads one.
+    pub fn read(&mut self, addr: DAC_ADDR) -> u32 {
+        let mut buf = [0u8; 8];
+        buf[0] = 1u8 << 7 | (addr as u32 >> 24) as u8;
+        self.sync_n.set_low();
+        self.spi.transfer(&mut buf).unwrap();
```

Comment:
You probably want to deassert SYNC in between. That then also makes the payload data conversion a bit easier and give more value to you already having shifted the address 24 bit left in the enum.

![image](https://user-images.githubusercontent.com/1338946/191211793-e490fbcb-4943-45c6-9433-76bc8648e985.png)

Reply:
Ok. I use a 1 us delay for the 100 ns SYNC high timing now. Since we currently only need to read during the DAC initialization we can use the `setup()` `delay` that implements `delay_us()`.